src/ImageSharp/Formats/Jpeg/Components/ZigZag.Intrinsic.cs

```diff
@@ -225,69 +239,61 @@ public static unsafe void ApplyZigZagOrderingSsse3(ref Block8x8 block)
         /// Applies zig zag ordering for given 8x8 matrix using AVX cpu intrinsics.
         /// </summary>
         /// <param name="block">Input matrix.</param>
-        public static unsafe void ApplyZigZagOrderingAvx2(ref Block8x8 block)
+        public static unsafe void ApplyTransposingZigZagOrderingAvx2(ref Block8x8 block)
         {
             DebugGuard.IsTrue(Avx2.IsSupported, "Avx2 support is required to run this operation!");
 
             fixed (byte* shuffleVectorsPtr = AvxShuffleMasks)
             {
-                Vector256<byte> rowsAB = block.V01.AsByte();
-                Vector256<byte> rowsCD = block.V23.AsByte();
-                Vector256<byte> rowsEF = block.V45.AsByte();
-                Vector256<byte> rowsGH = block.V67.AsByte();
-
-                // rows 0 1
-                Vector256<int> rows_AB01_EF01_CD23_shuffleMask = Avx.LoadVector256(shuffleVectorsPtr + (0 * 32)).AsInt32();
-                Vector256<byte> row01_AB = Avx2.PermuteVar8x32(rowsAB.AsInt32(), rows_AB01_EF01_CD23_shuffleMask).AsByte();
+                Vector256<byte> rowAB = block.V01.AsByte();
+                Vector256<byte> rowCD = block.V23.AsByte();
+                Vector256<byte> rowEF = block.V45.AsByte();
+                Vector256<byte> rowGH = block.V67.AsByte();
+
+                /* row01 - A0 B0 A1 A2 B1 C0 D0 C1 | B2 A3 A4 B3 C2 D1 E0 F0 */
+                Vector256<int> crln_01_AB_CD = Avx.LoadVector256(shuffleVectorsPtr + (0 * 32)).AsInt32();
+                Vector256<byte> row01_AB = Avx2.PermuteVar8x32(rowAB.AsInt32(), crln_01_AB_CD).AsByte();
```

Comment:
Any chance we can group Loads and PermuteVars along the implementation (without running out of registers) so they are pipelined? 

See https://github.com/SixLabors/ImageSharp/pull/1242#issuecomment-735346693

Reply:
Cross-lane permutations are not pipelined:
![image](https://user-images.githubusercontent.com/20967409/151132068-44d07570-9250-4e76-8ec6-b723bdeea8c0.png)

Loads are the main bottlenecks here with latency of **7** which can be pipelined:
![image](https://user-images.githubusercontent.com/20967409/151132240-7b598fe3-ce58-408e-a97a-20794caeffc1.png)

Even though shuffles and permutes don't pipeline well, shuffles can be pipelined on the newer architecture:
![image](https://user-images.githubusercontent.com/20967409/151132510-c24c1124-4db0-426c-90c6-e63750cb351f.png)

So I'll try to group them without register spilling as well but I don't have such CPU to test performance.

Reply:
Nah, grouped loads code is a bit slower:

``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
Intel Core i7-6700K CPU 4.00GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=6.0.100-preview.3.21202.5
  [Host]     : .NET Core 3.1.21 (CoreCLR 4.700.21.51404, CoreFX 4.700.21.51508), X64 RyuJIT
  DefaultJob : .NET Core 3.1.21 (CoreCLR 4.700.21.51404, CoreFX 4.700.21.51508), X64 RyuJIT


```
|          Method |     Mean |     Error |    StdDev | Code Size |
|---------------- |---------:|----------:|----------:|----------:|
| &#39;Avx optimized&#39; | 5.666 ns | 0.1440 ns | 0.2018 ns |     478 B |
| &#39;Avx pipelined&#39; | 6.076 ns | 0.1516 ns | 0.2175 ns |     501 B |

While loads are grouped at the assembler level, 'pipelined' version saves 6 `xmm` (yes, `xmm`, not `ymm`) registers to the stack while current PR versions saves 4 registers to the stack for whatever reason, thus the slowdown.

Reply:
> Cross-lane permutations are not pipelined:

To me Latency = 3, Througput = 1 means that you can pipeline 3 instructions so the overall latency will be 3 cycles instead of 3 * 3 cycles, which is already a win.

> While loads are grouped at the assembler level, 'pipelined' version saves 6 `xmm` (yes, `xmm`, not `ymm`) registers to the stack while current PR version saves 4 registers to the stack for whatever reason, thus the slowdown.

That's super weird, especially the xmm part.

Anyways, this was just an idea, the improvement in the PR is already good enough.

Reply:
It's the opposite: latency is how many 'cycles' single independent op takes, throughput is how many instructions can be started in one 'cycle', it's a reciprocal value though.

Let's take unaligned loads for example, latency of 7 and thoughput 0.5 on newer architecture, let's say we have this code:
```asm
...
vmovupd   ymm0,[rcx]
vmovupd   ymm1,[rcx+20]
vmovupd   ymm2,[rcx+40]
vmovupd   ymm3,[rcx+60]
...
```

Due to the fancy cpu out-of-order first two `vmovupd` loads can be issued simultaneously because they are independent. In the best case possible code above would be executed in 8 cycles:

```
|1 cycle|2 cycle|3 cycle|4 cycle|5 cycle|6 cycle|7 cycle|8 cycle    |
|vmovupd                                                |use results|
|vmovupd                                                |use results|
        |vmovupd                                                    |use results|
        |vmovupd                                                    |use results|
```

This is a very rough 'view' of what could happen because all these `vmovupd` must have been decoded and present in instruction queue, there should be no other port competing instructions there and other complicated stuff. But general idea is this.

Reply:
I think you have pipelining confused with parallel dispatch.

A pipelined instruction is one that takes more than one cycle to complete but does not tie up the execution unit completely during that time, so instructions can overlap execution within that latency window.  You've correctly illustrated that for `vmovupd` above, where both pipelining and parallel dispatch apply, because 2 ports can accept that instruction simultaneously and up to 7 instructions can be in various stages of completion within the EU's pipeline.

In the case of `vpermd`, the latency is 3, meaning it takes 3 cycles from dispatch to completion, and its reciprocal throughput is 1, meaning an instruction can be dispatched and retired every cycle (this one can only be dispatched on a single port).  The total execution time for 2 `vpermd`s would be 4 cycles, with both results available on the 5th.

```
|1 cycle|2 cycle|3 cycle|4 cycle    |5 cycle    |
|vpermd                 |use results|
        |vpermd                     |use results|
```

So there is potential benefit in unrolling those.

Reply:
I actually did, thanks for clarifying!
But as I said, grouped loads implementation is slower. Maybe grouped permutations and shuffles would yield better performance with enough manual tweaking and looking bad at the binary but with 'naive' grouping them  in code yields worse/same results:

|                                           Method |     Mean |     Error |    StdDev | Code Size |
|------------------------------------------------- |---------:|----------:|----------:|----------:|
|                                  'Avx detailed masks' | 9.309 ns | 0.0382 ns | 0.0319 ns |     569 B |
|                                  'Avx optimized' | 5.965 ns | 0.0696 ns | 0.0617 ns |     478 B |
|                            'Avx pipelined loads' | 6.095 ns | 0.0346 ns | 0.0307 ns |     501 B |
|                    'Avx pipelined the right way' | 6.424 ns | 0.0314 ns | 0.0278 ns |     488 B |

Legend:
Detailed masks - first implementation with detailed permutation/shuffle masks
Optimized - reduced permutation/shuffle masks table
Pipelined loads - grouped mask loads
The right way - grouped everything manually to get the best codegen from the 'pipeline loads' implementation

For example, here's the codegen for the first 2 rows calculation:

```asm
...
vmovupd   ymm0,[rcx]
vmovupd   ymm1,[rcx+20]
vmovupd   ymm2,[rcx+40]
vmovdqu   ymm3,ymmword ptr [rax]
vmovdqu   ymm4,ymmword ptr [rax+20]
vmovdqu   ymm5,ymmword ptr [rax+40]
vmovdqu   ymm6,ymmword ptr [rax+60]
vmovdqu   ymm7,ymmword ptr [rax+80]
vpermd    ymm8,ymm3,ymm0
vpermd    ymm3,ymm3,ymm1
vpermd    ymm9,ymm6,ymm2
vpshufb   ymm8,ymm8,ymm4
vpshufb   ymm3,ymm3,ymm5
vpshufb   ymm4,ymm9,ymm7
vpor      ymm3,ymm8,ymm3
vpor      ymm3,ymm3,ymm4
vmovupd   [rdx],ymm3
...
```

But yet this produces same performance (if not worse, benchmark is a bit unstable for whatever reason). Maybe I'm doing something wrong, I can provide full code for all of those benchmark versions if you want to play with this code but I'm honestly okay with current performance :)

Reply:
Yep, that codegen looks good.  I'd be happy to look at your variants if you want to share them, but if you're already happy with the perf, I'm in no position to argue ;)

This is something that may be worth revisiting once netcoreapp3.1 is EOL.  The constant vector and vector CSE support in net6.0 might help.  It's difficult to get optimal code on both runtimes at present.

Reply:
>I'm honestly okay with current performance

As am I 😄 

This has been a fantastic discussion btw, I've learned a lot!

Reply:
I'll leave it here just in case you'd want to take a look:
https://gist.github.com/br3aker/4da983868b8915180747ba1f8274862e